### PR TITLE
[server] Delete webhook events directly

### PR DESCRIPTION
**Pull request by easyCZ**
## Related Issue(s)

Fixes #

#### Preview status

gitpod:summary

## Build Options:

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [x] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

/hold

### CODE REVIEW

File: components/gitpod-db/src/typeorm/webhook-event-db-impl.ts [hunk 1]
```diff
@@ -52,19 +52,17 @@ export class WebhookEventDBImpl implements WebhookEventDB {
...
-            .update()
-            .set({ deleted: true })
-            .where("creationTime <= :expirationDate", { expirationDate });
-        if (typeof limit === "number") {
-            query.limit(limit);
-        }
-        await query.execute();
+            .limit(limit)
+            .where("creationTime <= :expirationDate", { expirationDate })
+            .delete();
```
Comment (easyCZ) [inline components/gitpod-db/src/typeorm/webhook-event-db-impl.ts:64]: This was previously reverted in https://github.com/gitpod-io/gitpod/commit/e0f8cd3251dae5003d71fe2ca898387900da3d92

The problem was the extra `execute` statement, which is now removed.

Comment (svenefftinge) [reply components/gitpod-db/src/typeorm/webhook-event-db-impl.ts:64]: IIRC we also found out that limit is ignored in delete statements, no?

Comment (svenefftinge) [reply components/gitpod-db/src/typeorm/webhook-event-db-impl.ts:64]: alternative could be to delete with a limited sub select:

```
DELETE t.* FROM table_name t 
JOIN (
  SELECT id FROM table_name 
  ORDER BY id 
  LIMIT 10
) t1 ON t.id = t1.id;
```

Comment (geropl) [reply components/gitpod-db/src/typeorm/webhook-event-db-impl.ts:64]: DELETE ... LIMIT should just work, at least in MySQL world. Not sure if typeorm breaks stuff, though...

Comment (svenefftinge) [reply components/gitpod-db/src/typeorm/webhook-event-db-impl.ts]: Oh sorry for spreading FUD here. It should indeed work for such simple delete statements. It just doesn't work in DELETES with joins 
> You cannot use ORDER BY or LIMIT in a multiple-table DELETE.
(see https://dev.mysql.com/doc/refman/5.7/en/delete.html)

**Review by svenefftinge — APPROVED** [review]

File: components/gitpod-db/src/typeorm/webhook-event-db-impl.ts (hunk 1 above)
Comment (easyCZ) [reply components/gitpod-db/src/typeorm/webhook-event-db-impl.ts]: I'm fighting dbtests in my workspace a bit, trying to validate the exact query it uses for this. Will get back here once I have that.

Comment (easyCZ) [reply components/gitpod-db/src/typeorm/webhook-event-db-impl.ts]: Looks like it indeed ignores the limit. It executes 
```
DELETE FROM `d_b_webhook_event` WHERE `creationTime` <= :expirationDate
```

Comment (easyCZ) [reply components/gitpod-db/src/typeorm/webhook-event-db-impl.ts]: I've rewritten the query

File: components/gitpod-db/src/typeorm/webhook-event-db-impl.ts [hunk 2]
```diff
@@ -52,19 +52,23 @@ export class WebhookEventDBImpl implements WebhookEventDB {
...
-            .set({ deleted: true })
-            .where("creationTime <= :expirationDate", { expirationDate });
-        if (typeof limit === "number") {
-            query.limit(limit);
-        }
-        await query.execute();
+        await repo.query(
+            `DELETE FROM d_b_webhook_event
+                WHERE id IN (
+                SELECT id FROM (
```
Comment (geropl) [inline components/gitpod-db/src/typeorm/webhook-event-db-impl.ts:63]: The subselect feels overly complicated. Is [the problem we're seeing here](https://github.com/gitpod-io/gitpod/pull/17650#discussion_r1197578844) with Typeorm?

Then we should do `DELETE FROM d_b_webhook_event WHERE creationTime <= ? LIMIT ?` I think.
Am I missing something? :thinking:

Comment (easyCZ) [reply components/gitpod-db/src/typeorm/webhook-event-db-impl.ts:63]: It was based on this https://stackoverflow.com/questions/7142097/mysql-delete-statement-with-limit

Comment (geropl) [reply components/gitpod-db/src/typeorm/webhook-event-db-impl.ts:63]: That question is about offsets (` LIMIT 20, 50`).

Comment (easyCZ) [reply components/gitpod-db/src/typeorm/webhook-event-db-impl.ts]: updated

**Review by geropl — APPROVED** [review]
:heavy_check_mark:

**Pull request merged**